File: app/schemas/agent.py
from datetime import datetime
from typing import Any

from .base import BaseModel


class AgentSessionCreateRequest(BaseModel):
    title: str | None = None
    model_id: int | None = None
    document_id: int | None = None
    section_id: int | None = None
    enable_mcp: bool = False


class AgentSessionUpdateRequest(BaseModel):
    title: str | None = None
    model_id: int | None = None
    thinking_level: str | None = None  # off | low | medium | high
    permission_mode: str | None = None  # manual | auto | bypass
    auto_allow_tools: list[str] | None = None
    enable_mcp: bool | None = None


class AgentSessionInfo(BaseModel):
    id: int
    uuid: str
    title: str
    status: str
    adapter: str
    model_id: int | None = None
    document_id: int | None = None
    section_id: int | None = None
    enable_mcp: bool
    thinking_level: str
    permission_mode: str
    auto_allow_tools: list[str] | None = None
    context: dict[str, Any] | None = None
    create_time: datetime
    update_time: datetime | None = None


class AgentReferenceIn(BaseModel):
    """用户在输入框里 `@` 出来的一个对象。

    **正文里只有标题,id 走这里。** 标题会重、会改、会带空格,拿它当标识迟早出事;
    把数字 id 塞回正文又会把真正的句子挤没。所以句子照旧是人话(「把 @周报 里的三点
    合成一段摘要」),而这份结构化清单负责告诉模型那三个字究竟指向哪一条。
    """

    kind: str
    id: int
    name: str = ""


class AgentMessageCreateRequest(BaseModel):
    content: str
    images: list[str] = []
    references: list[AgentReferenceIn] = []


class AgentMessageInfo(BaseModel):
    id: int
    session_id: int
    role: str
    content: str
    payload: dict[str, Any] | None = None
    error: str | None = None
    create_time: datetime


class AgentToolInvokeRequest(BaseModel):
    arguments: dict[str, Any] = {}
    # 确认卡上显示的请求方;留空用默认("pi-agent")。
    # **没有 session_id**:这次调用属于哪次对话,由调用方的令牌说了算 —— 参数说的可以是任何值。
    requested_by: str = ""


class AgentToolSpecInfo(BaseModel):
    name: str
    description: str
    parameters: dict[str, Any]
    confirmation: bool = False
    permission: str | None = None
    read_only: bool = False


class ToolConfirmationInfo(BaseModel):
    id: int
    session_id: int | None = None
    tool: str
    permission: str
    summary: str
    payload: dict[str, Any]
    status: str
    # 工具的返回什么形状都有(dict/str/list)。**不要写成 `Any | None`**:那会让
    # OpenAPI 产出 anyOf:[{}, null],openapi-generator 会生成一个引用不到的空壳类型。
    result: Any = None
    error: str | None = None
    requested_by: str
    decided_by: int | None = None
    create_time: datetime
    resolved_at: datetime | None = None


class AgentCompactResponse(BaseModel):
    context: dict[str, Any] | None = None
    compaction: dict[str, Any] | None = None

File: web/src/types/agent.ts
/**
 * Revornix AI(智能体)的前端类型。与 api/router/agent.py + agent/host.py 的形状一一对应。
 */

export type AgentSession = {
	id: number;
	uuid: string;
	title: string;
	status: 'idle' | 'running';
	adapter: string;
	model_id: number | null;
	document_id: number | null;
	section_id: number | null;
	enable_mcp: boolean;
	thinking_level: 'off' | 'low' | 'medium' | 'high';
	permission_mode: 'manual' | 'auto' | 'bypass';
	auto_allow_tools: string[];
	context?: { tokens: number; window: number } | null;
	create_time: string;
	update_time?: string | null;
};

/**
 * `@` 引用得到的东西:知识库里一个具体对象,不是一段字。
 *
 * 类型放在这里而不是组件旁边,是因为它同时出现在**三条路**上:输入框的文档、发出去的
 * 请求体、落库后的消息 payload。搜索与图标那套东西留在 components/revornixai/references.ts。
 */
export type ReferenceKind = 'document' | 'section';

export type AgentReference = {
	kind: ReferenceKind;
	id: number;
	/** 写进正文的那个标题。存下来是为了旧消息也能渲染。 */
	name: string;
};

/** 时间线条目:一轮回答里发生的事,按真实顺序。 */
export type AgentTimelineItem =
	| { type: 'text'; text: string }
	| { type: 'thinking'; text: string; done?: boolean }
	| { type: 'tool'; tool: AgentToolCall }
	| { type: 'subtool'; parent_id: string; tool: AgentToolCall };

export type AgentToolCall = {
	id: string;
	name?: string;
	args?: unknown;
	status: 'running' | 'done' | 'error';
	result?: unknown;
	usage?: {
		started_at?: string;
		finished_at?: string;
		duration_seconds?: number;
	};
};

export type AgentCitation = {
	document_id: number;
	document_title: string;
	chunk_id: string;
	excerpt: string;
	score?: number | null;
};

export type AgentMessagePayload = {
	usage?: {
		duration_seconds?: number;
		first_token_seconds?: number;
		metering?: Record<string, unknown>;
	};
	context?: { tokens: number; window: number };
	compaction?: {
		droppedMessages: number;
		tokensBefore: number;
		tokensAfter: number;
		summary: string;
	};
	citations?: AgentCitation[];
	timeline?: AgentTimelineItem[];
	queued?: boolean;
	images?: string[];
	/** 用户在输入框里 `@` 出来的对象。**随消息存下来**,所以旧气泡也画得出胶囊 ——
	 *  被引用的东西改了名或删了,至少还说得出当时指的是什么。 */
	references?: AgentReference[];
};

export type AgentMessage = {
	id: number;
	session_id: number;
	role: 'user' | 'assistant' | 'system';
	content: string;
	payload?: AgentMessagePayload | null;
	error?: string | null;
	create_time: string;
};

export type ToolConfirmation = {
	id: number;
	session_id: number | null;
	tool: string;
	permission: 'edit' | 'ai-cost' | 'destructive';
	summary: string;
	payload: Record<string, unknown>;
	status: 'pending' | 'approved' | 'executed' | 'rejected' | 'failed' | 'cancelled';
	result?: unknown;
	error?: string | null;
	requested_by: string;
	decided_by?: number | null;
	create_time: string;
	resolved_at?: string | null;
};

/** SSE 流帧(快照式):GET /agent/session/{id}/stream */
export type AgentStreamFrame = {
	text: string;
	done: boolean;
	timeline: AgentTimelineItem[];
};
